### PR TITLE
Fix truncated hex in anvil dump_state

### PR DESCRIPTION
**Pull request by AxelAramburu**
## Motivation

Fixes https://github.com/foundry-rs/foundry/issues/8179

An issue reported is leading zeros in hex can be truncated and leading to deserialize failure with `anvil_dumpState`

## Solution

Like @mattsse recommend in the issue post, I change the type of storage from `U256` to `B256`.
https://github.com/foundry-rs/foundry/blob/f6730662831d0473f7207f4c0474a4bb6ee016eb/crates/anvil/src/eth/backend/db.rs#L345-L345 

The modifications don't break the actual implemented tests:

Before
```
...

"0x068e44eb31e111028c41598e4535be7468674d0a": { // line 32, a contract address

...

"0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc": "0x68e44eb31e111028c41598e4535be7468674d0a", // line 172, a reference to that contract
"0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103": "0x9e2803ad9ecc0a848c8a045329524d295f6ee44f",
"0xd30e835d3f35624761057ff5b27d558f97bd5be034621e62240e5c0b784abe68": "0x9965507d1a55bcc2695c58ba16fb37d819b0a4dc"
```

After
```
...

"0x068e44eb31e111028c41598e4535be7468674d0a": { // line 32, a contract address

...

"0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc": "0x000000000000000000000000068e44eb31e111028c41598e4535be7468674d0a", // <--- truncated zero is preserved
"0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103": "0x0000000000000000000000009e2803ad9ecc0a848c8a045329524d295f6ee44f",
"0xd30e835d3f35624761057ff5b27d558f97bd5be034621e62240e5c0b784abe68": "0x0000000000000000000000009965507d1a55bcc2695c58ba16fb37d819b0a4dc"
```

Let me know if I can improve my contribution

### CODE REVIEW

File: crates/anvil/src/eth/backend/db.rs [hunk 1]
```diff
@@ -348,7 +348,7 @@ pub struct SerializableAccountRecord {
     pub nonce: u64,
     pub balance: U256,
     pub code: Bytes,
-    pub storage: BTreeMap<U256, U256>,
+    pub storage: BTreeMap<U256, B256>,
```
Comment (mattsse) [inline crates/anvil/src/eth/backend/db.rs:351]: this should be flipped, the key should always be b256

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/anvil/src/eth/backend/db.rs (hunk 1 above)
Comment (AxelAramburu) [reply crates/anvil/src/eth/backend/db.rs]: I change also the key type

Comment (DaniPopes) [reply crates/anvil/src/eth/backend/db.rs]: That's good, but the value should stay U256

Comment (AxelAramburu) [reply crates/anvil/src/eth/backend/db.rs]: If I the value is U256 the bug is not corrected:
https://github.com/foundry-rs/foundry/issues/8179
```
{
"0x0000000000000000000000000000000000000000000000000000000000000000":"0x1",
"0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc":"0x68e44eb31e111028c41598e4535be7468674d0a", // <--- The truncated zero
"0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103":"0x9e2803ad9ecc0a848c8a045329524d295f6ee44f",
"0xd30e835d3f35624761057ff5b27d558f97bd5be034621e62240e5c0b784abe68":"0x9965507d1a55bcc2695c58ba16fb37d819b0a4dc"}},
```

File: crates/anvil/src/eth/backend/db.rs [hunk 2]
```diff
@@ -348,7 +348,7 @@ pub struct SerializableAccountRecord {
     pub nonce: u64,
     pub balance: U256,
     pub code: Bytes,
-    pub storage: BTreeMap<U256, U256>,
+    pub storage: BTreeMap<B256, B256>,
```
Comment (mattsse) [inline crates/anvil/src/eth/backend/db.rs:351]: this is also how the regular genesis account is formatted with hash -> hash

File: crates/anvil/src/eth/backend/mem/in_memory_db.rs
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/in_memory_db.rs:1]: please cargo +nightly fmt

File: crates/anvil/src/eth/backend/mem/mod.rs [hunk 3]
```diff
@@ -564,7 +564,7 @@ impl Backend {
         slot: U256,
         val: B256,
     ) -> DatabaseResult<()> {
-        self.db.write().await.set_storage_at(address, slot, U256::from_be_bytes(val.0))
+        self.db.write().await.set_storage_at(address, B256::from(slot), B256::from(val.0))
```
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/mod.rs:567]: `B256::from(val.0)` is not needed here?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/anvil/src/eth/backend/mem/in_memory_db.rs
Comment (AxelAramburu) [reply crates/anvil/src/eth/backend/mem/in_memory_db.rs:1]: It's good

File: crates/anvil/src/eth/backend/mem/mod.rs (hunk 3 above)
Comment (AxelAramburu) [reply crates/anvil/src/eth/backend/mem/mod.rs]: I fixed it thanks

File: crates/anvil/src/eth/backend/db.rs (hunk 2 above)
Comment (AxelAramburu) [reply crates/anvil/src/eth/backend/db.rs:351]: I see an example of genesis.json storage field: 

```
      "storage": {
        "0x0000000000000000000000000000000000000000000000000000000000000001": "0x02",
        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x04",
        "0x29ecdbdf95c7f6ceec92d6150c697aa14abeb0f8595dd58d808842ea237d8494": "0x01",
        "0x6aa118c6537572d8b515a9f9154be55a3377a8de7991cd23bf6e5ceb368688e3": "0x01",
        ...
      }
```

The type is `pub storage: BTreeMap<B256, U256>`, but it doesnt resolve the issue:
https://github.com/foundry-rs/foundry/issues/8179

File: crates/anvil/src/eth/backend/mem/in_memory_db.rs
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/in_memory_db.rs:1]: something went wrong here, why are all lines changed?
line endings?

**Comment by zerosnacks** [comment]
Hi @AxelAramburu why did you close your PR? Do you need any help / pointers? Would be great to get this fix in

**Comment by KholdStare** [comment]
@AxelAramburu Running into this issue. Can you re-open?

**Comment by 0xalpharush** [comment]
What remains to be done here? I am also having issues decoding the state dump due to odd-length fields which are expected to be hex encoded

File: crates/anvil/src/eth/backend/mem/mod.rs
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/mod.rs:1]: looks like this is actually unchanged?

perhaps some whitespace issue idk

**Review by mattsse — COMMENTED** [review]
actual changes look good, can we undo the changes in backend/mem/mod.rs @zerosnacks then this should be good

**Comment by 0xalpharush** [comment]
I don't think this fixes the issue entirely.
Run `foundryup -r AxelAramburu/foundry -b fix_hex_truncated`
```
$ anvil --version             
anvil 0.2.0 (8326b48e 2024-09-12T16:47:11.876055000Z)
```
In one terminal:
```
$ anvil --dump-state test.json
```
In a second (script from template produced by `foundry init`):
```
forge script script/Counter.s.sol --rpc-url 127.0.0.1:8545 --broadcast -vvvv --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
```

Still has odd-length
```
{
    "block": {
        "number": "0x1",
        "coinbase": "0x0000000000000000000000000000000000000000",
        "timestamp": "0x66e31db6",
        "gas_limit": "0x1c9c380",
        "basefee": "0x3b9aca00",
        "difficulty": "0x0",
        "prevrandao": "0x481952e621a89f45b1d7520a5729949d946c688b21bfad7e578a5d9b2cc30f27",
        "blob_excess_gas_and_price": {
            "excess_blob_gas": 0,
            "blob_gasprice": 0
        }
    },
    "accounts": {
        "0x0000000000000000000000000000000000000000": {
            "nonce": 0,
            "balance": "0x1a0ed",
            "code": "0x",
            "storage": {}
        },
        "0x14dc79964da2c08b23698b3d3cc7ca32193d9955": {
            "nonce": 0,
            "balance": "0x21e19e0c9bab2400000",
            "code": "0x",
            "storage": {}
        },
        "0x15d34aaf54267db7d7c367839aaf71a00a2c6a65": {
            "nonce": 0,
            "balance": "0x21e19e0c9bab2400000",
            "code": "0x",
            "storage": {}
        },
        "0x23618e81e3f5cdf7f54c3d65f7fbc0abf5b21e8f": {
            "nonce": 0,
            "balance": "0x21e19e0c9bab2400000",
            "code": "0x",
            "storage": {}
        },
        "0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc": {
            "nonce": 0,
            "balance": "0x21e19e0c9bab2400000",
            "code": "0x",
            "storage": {}
        },
        "0x4e59b44847b379578588920ca78fbf26c0b4956c": {
            "nonce": 0,
            "balance": "0x0",
            "code": "0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe03601600081602082378035828234f58015156039578182fd5b8082525050506014600cf3",
            "storage": {}
        },
        "0x5fbdb2315678afecb367f032d93f642f64180aa3": {
            "nonce": 1,
            "balance": "0x0",
            "code": "0x6080604052348015600f57600080fd5b5060043610603c5760003560e01c80633fb5c1cb1460415780638381f58a146053578063d09de08a14606d575b600080fd5b6051604c3660046083565b600055565b005b605b60005481565b60405190815260200160405180910390f35b6051600080549080607c83609b565b9190505550565b600060208284031215609457600080fd5b5035919050565b60006001820160ba57634e487b7160e01b600052601160045260246000fd5b506001019056fea2646970667358221220d6eead0a2debb1972173eae3e5f1636066a01ff9f3a14fe42618a17daa0af59264736f6c634300081a0033",
            "storage": {}
        },
        "0x70997970c51812dc3a010c7d01b50e0d17dc79c8": {
            "nonce": 0,
            "balance": "0x21e19e0c9bab2400000",
            "code": "0x",
            "storage": {}
        },
        "0x90f79bf6eb2c4f870365e785982e1f101e93b906": {
            "nonce": 0,
            "balance": "0x21e19e0c9bab2400000",
            "code": "0x",
            "storage": {}
        },
        "0x976ea74026e726554db657fa54763abd0c3a0aa9": {
            "nonce": 0,
            "balance": "0x21e19e0c9bab2400000",
            "code": "0x",
            "storage": {}
        },
        "0x9965507d1a55bcc2695c58ba16fb37d819b0a4dc": {
            "nonce": 0,
            "balance": "0x21e19e0c9bab2400000",
            "code": "0x",
            "storage": {}
        },
        "0xa0ee7a142d267c1f36714e4a8f75612f20a79720": {
            "nonce": 0,
            "balance": "0x21e19e0c9bab2400000",
            "code": "0x",
            "storage": {}
        },
        "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266": {
            "nonce": 1,
            "balance": "0x21e19e068a7fbb15d13",
            "code": "0x",
            "storage": {}
        }
    },
    ```

**Comment by KholdStare** [comment]
Maybe this is wrong, but I think there should be "round-trip" test in general for the dump state functionality. Given a state, does dump_state -> load_state -> dump_state produce the same result as the first dump_state? It should catch any random regressions that may happen in the future.

**Review by mattsse — APPROVED** [review]
good point, we should be able to easily add one

@yash-atreya could you please add such a test? but we can also do this as a followup

**Pull request merged**